core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -427,10 +386,10 @@ void toSerializableWithExtraRead(DistributedStorage storage)
             Scanner scanner = null;
             try {
               Scan scan = entry.getKey();
-              // only get tx_id and tx_version columns because we use only them to compare
-              scan.clearProjections();
-              scan.withProjection(Attribute.ID).withProjection(Attribute.VERSION);
-              ScalarDbUtils.addProjectionsForKeys(scan, getTableMetadata(scan));
+              // For JDBC databases or non-JDBC databases without filtering, we only get tx_id,
+              // tx_version, and primary key columns because we use only them to compare. For
+              // non-JDBC databases with filtering, we also need filtering columns to compare.
+              projectRequiredColumns(storage, scan);
```

Comment:
This is the fix for a bug with the serializable consensus commit. We need non-primary key columns to perform filtering correctly in the ScalarDB layer if they are used in the conditions, even for the extra-read purpose.

Reply:
[minor] Does it mean that the values of `scanSet` doesn't contain filtering columns in the case of JDBC (thanks to the condition pushdown?), while the values of `scanSet` contain filtering columns in the case of non-JDBC?

If so, I guess it would be a bit more helpful to understand if `Scan` tells whether filtering columns are contained with a method like `scan.containsFilteringColumnsInKey()` so that we can revise the condition in `projectRequiredColumns` like this?

```java
if (scan.containsFilteringColumnsInKey()) {
  scan.getConjunctions()
      .forEach(
          conjunction ->
              conjunction
   ...
```

Reply:
I might not get your point perfectly, but it doesn't mean whether the user's original `Scan` has filtering columns in projections or not. Do you think this condition `if (!(storage instanceof JdbcDatabase))` is hard to understand the background? For JDBC databases, we don't have to care about the existence of filtering columns in projections, so I didn't want to check it excessively.

By the way, thanks to your question, I might find an issue regarding projections...

Reply:
Thanks for the reply!

What I thought was as follows:
- Why filtering columns were needed for non-JDBC (w/ filtering) ?
- The results of this scan are put to `currentReadMap` as a part of the key
- Then, the map is looked-up with the values of `scanSet` (`TransactionResult latestResult = currentReadMap.get(key)`)
- So, including filtering columns in projection is for consistency of the key types related to the `equals()`'s logic?

But I think I had the wrong idea... Can you elaborate a bit more why filtering columns are needed for projection in the case of non-JDBC with filtering?

Reply:
Ahh, the reason is not related to the key consistency of `currentReadMap` since the validation is always handled with the primary key basis.

We originally project columns except for `tx_id`, `tx_version`, and primary key columns for efficiency since they are enough for the re-scan for the validation phase (extra-read). We now want to filter results in the ScalarDB layer for non-JDBC databases. So, we need to additionally have filtering columns here since they are used even for the re-scan, the same as the first scan.

Reply:
Yeah, I misread the code around Key class... Sorry for bothering you.

> We originally project columns except for tx_id, tx_version, and primary key columns

`s/except for/only for/` ?

But, I think I don't still get why columns for filtering need to be projected. My understanding is as follows:
- `Scan` instances stored in `scanSet` contain the original conditions
- `FilterableScannerImpl#getResultWithFiltering` filters records based on the conditions. Whether the columns for filtering are projected isn't related to the filtering. (The columns for filtering included in the conditions are automatically scanned at the underlying storage layer, right?)
- The columns for filtering of the result in `currentReadMap` are not used
- So, the columns for filtering don't need to be projected?

I think I have some wrong ideas of them. Can you point out it?

Reply:
> Yeah, I misread the code around Key class... Sorry for bothering you.

No no, it's OK. I think this part is so complicated and worth while to share the background.

> > We originally project columns except for tx_id, tx_version, and primary key columns
> 
> `s/except for/only for/` ?

Sorry, "only for".

> But, I think I don't still get why columns for filtering need to be projected. My understanding is as follows:
> 
> * `Scan` instances stored in `scanSet` contain the original conditions

Right. Precisely, the list of projections is basically cleared in `CrudHandler` for the sake of transaction processing. In that sense, it's a bit different from the original one specified by users.

> * `FilterableScannerImpl#getResultWithFiltering` filters records based on the conditions. Whether the columns for filtering are projected isn't related to the filtering. (The columns for filtering included in the conditions are automatically scanned at the underlying storage layer, right?)

Right. We always need columns specified in the where clause to filter results. (*)

> * The columns for filtering of the result in `currentReadMap` are not used

Right. `currentReadMap` is for the read set verification. So, primary key columns are enough, but, of course, the map should be based on the result of the scan with the filtering conditions in the current database state.

> * So, the columns for filtering don't need to be projected?

No. We used them to get the current results above.

Let's chat a bit if you have further questions : )

Reply:
Thanks for the detailed answers!

I found my misunderstanding...

I thought `Cassandra.scan()` implicitly projects the columns for filtering when there are any conditions in the scan. What I expected was like this:

```java
--- a/core/src/main/java/com/scalar/db/storage/cassandra/Cassandra.java
+++ b/core/src/main/java/com/scalar/db/storage/cassandra/Cassandra.java
@@ -14,6 +14,7 @@ import com.scalar.db.api.Mutation;
 import com.scalar.db.api.Put;
 import com.scalar.db.api.Result;
 import com.scalar.db.api.Scan;
+import com.scalar.db.api.ScanBuilder.BuildableScanOrScanAllFromExisting;
 import com.scalar.db.api.Scanner;
 import com.scalar.db.common.AbstractDistributedStorage;
 import com.scalar.db.common.TableMetadataManager;
@@ -125,7 +126,9 @@ public class Cassandra extends AbstractDistributedStorage {
       return new ScannerImpl(handlers.select().handle(scan), resultInterpreter);
     } else {
       // Ignore limit to control it in FilterableScannerImpl
-      Scan scanAll = Scan.newBuilder(scan).limit(0).build();
+      BuildableScanOrScanAllFromExisting builder = Scan.newBuilder(scan);
+      withProjectionsForFilteringIfMissing(builder);
+      Scan scanAll = builder.limit(0).build();
       return new FilterableScannerImpl(scan, handlers.select().handle(scanAll), resultInterpreter);
     }
   }
```

BTW, assuming the above example works, it is a kind of provisioning the columns for filtering at storage layer while the change of this PR is provisioning the columns for filtering at an upper layer. Which do you think better?

Reply:
Right. As you thought, we needed to handle projections properly in the storage layer, and it would be better to hide such a storage matter from the upper layer. So, your question got to the core of the matter from the beginning. Thanks again. Fixed in 6b1e92a.